### PR TITLE
Move marble from api

### PR DESCRIPTION
**Pull request by Aherbeth**
- Correcting all API calls 
- Debugging and adapting client side

### CODE REVIEW

**Comment by Aherbeth** [comment]
The amount of file changed is due to the move of certain files from the client to the server side.

File: api/src/app.controller.ts [hunk 1]
```diff
@@ -25,14 +26,55 @@ export class AppController {
...
-  isMarblePlayable(): Boolean {
-    // return this.appService.getIsMarblePlayable();
-    return false;
+  @Get('restartgame')
+  getRestartGame(): GameState {
+    return this.appService.getRestartGame();
+  }
+
+  @Post('gamestatehaschanged')
+  getGameStateHasChanged(@Body() body): boolean {
```
Comment (djhi) [inline api/src/app.controller.ts:35]: This should probably be a `get`, not a `post`

File: api/src/app.controller.ts [hunk 2]
```diff
@@ -25,14 +26,55 @@ export class AppController {
...
+    return this.appService.gameStateHasChanged(body);
+  }
+
+  @Post('marbleplayable')
+  getMarblePlayable(@Body() body): Boolean {
```
Comment (djhi) [inline api/src/app.controller.ts:40]: Same. I know I told you `post` before but it seems I misunderstood the use case

File: api/src/app.service.ts [hunk 3]
```diff
@@ -14,6 +20,14 @@ export class AppService {
     return gameState;
   }
 
+  getRestartGame(): GameState {
```
Comment (djhi) [inline api/src/app.service.ts:23]:
```suggestion
  restartGame(): GameState {
```

File: api/src/app.service.ts [hunk 4]
```diff
@@ -14,6 +20,14 @@ export class AppService {
...
+    return createNewGameState();
+  }
+
+  gameStateHasChanged(playerGameState: GameState) {
```
Comment (djhi) [inline api/src/app.service.ts:27]:
```suggestion
  hasGameStateChanged(playerGameState: GameState) {
```

File: api/src/game.ts [hunk 5]
```diff
@@ -4,16 +4,25 @@ import { moveMarble } from './board';
 import { sanitizeGraph, boardToGraph } from './graph';
 import { INITIAL_BOARD } from './constants';
 
-export const gameState: GameState = {
+export let gameState: GameState = {
```
Comment (djhi) [inline api/src/game.ts:7]: Why ?

File: api/src/game.ts [hunk 6]
```diff
@@ -49,10 +59,8 @@ export const switchToNextPlayer = (
...
 ): Player => {
-  let nextPlayerIndex = players.indexOf(actualPlayer) + 1;
-
-  if (nextPlayerIndex > players.length - 1) return players[0];
-  return players[nextPlayerIndex];
+  if (actualPlayer.playerNumber === 1) return players[1];
+  return players[0];
 };
 
 export const marbleWonByPlayer = (graph: Graph): number => {
```
Comment (djhi) [inline api/src/game.ts:66]: I don't understand what this function is supposed to do and return. Does it make a marble be won by a player or does it check whether a marble has been won by a player?

File: api/src/app.service.ts [hunk 7]
```diff
@@ -47,6 +65,7 @@ export class AppService {
       gameState.currentPlayer,
       gameState.players,
     );
+    setGameState(newGameState);
```
Comment (djhi) [inline api/src/app.service.ts:68]: As `setGameState` return the new state, I would do that like this:

```js
return setGameState({
    ....gameState,
    graph: moveMarbleInDirection(graph, coordinates, direction),
    currentPlayer: switchToNextPlayer(
      gameState.currentPlayer,
      gameState.players,
    );
});
```

File: api/src/game.ts [hunk 8]
```diff
@@ -108,3 +116,8 @@ export const setGameStateOnDirectionSelected = (
     gameState.players,
   );
 };
+
+export const setGameState = (newGameState: GameState): GameState => {
+  gameState = { ...newGameState };
```
Comment (djhi) [inline api/src/game.ts:121]: Why the spread? For cloning?

File: client/src/apiCalls.ts [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    return gameState;
+  } catch (ex) {
+    throw new GameError("The game can't be launched");
+  }
+};
+
+export const pullGameState = async (): Promise<GameState> => {
+  try {
+    const response = await fetch(`${URL}/gamestate`);
+    const gameState: GameState = (await response.json()) as GameState;
```
Comment (djhi) [inline client/src/apiCalls.ts:49]: I would put the casting at the next line to make this line more readable

File: client/src/apiCalls.ts [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+export const pullGameStateChanged = async (
+  playerGameState: GameState,
+): Promise<boolean> => {
+  try {
+    const response = await fetch(`${URL}/gamestatehaschanged`, {
+      method: 'POST',
+      body: JSON.stringify(playerGameState),
+      headers: { 'Content-Type': 'application/json' },
+    });
+    const gameStateHasChanged = (await response.json()) as boolean;
```
Comment (djhi) [inline client/src/apiCalls.ts:65]: Same

File: client/src/apiCalls.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  direction: string,
+  player: Player,
+): Promise<boolean> => {
+  try {
+    const response = await fetch(`${URL}/marbleplayable`, {
+      method: 'POST',
+      body: JSON.stringify({ gameState, direction, player }),
+      headers: { 'Content-Type': 'application/json' },
+    });
+    const canMove: boolean = (await response.json()) as boolean;
```
Comment (djhi) [inline client/src/apiCalls.ts:83]: Same

File: client/src/apiCalls.ts [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+  player: Player,
+): Promise<GameState> => {
+  try {
+    const response = await fetch(`${URL}/movemarble`, {
+      method: 'POST',
+      body: JSON.stringify({ gameState, direction, player }),
+      headers: { 'Content-Type': 'application/json' },
+    });
+
+    const gameStateAfterMove: GameState = (await response.json()) as GameState;
```
Comment (djhi) [inline client/src/apiCalls.ts:103]: Same

File: client/src/apiCalls.ts [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+export const postGameState = async (
+  gameState: GameState,
+): Promise<GameState> => {
+  try {
+    const response = await fetch(`${URL}/setgamestate`, {
+      method: 'POST',
+      body: JSON.stringify(gameState),
+      headers: { 'Content-Type': 'application/json' },
+    });
+    const newGameState: GameState = (await response.json()) as GameState;
```
Comment (djhi) [inline client/src/apiCalls.ts:144]: Same

File: client/src/apiCalls.ts [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+};
+
+export const setCurrentState = (gameState: GameState) => {
+  currentState = gameState;
+};
+
+export const restartGame = async (): Promise<GameState> => {
+  try {
+    const response = await fetch(`${URL}/restartgame`);
+    const newGameState: GameState = (await response.json()) as GameState;
```
Comment (djhi) [inline client/src/apiCalls.ts:159]: Same

**Review by djhi — CHANGES_REQUESTED** [review]
I think you should rename the file `apiCalls.ts` to a simpler `api.ts`

File: api/src/app.controller.ts (hunk 1 above)
Comment (Aherbeth) [reply api/src/app.controller.ts:35]: That's what i thought at first, thanks :)

File: api/src/app.controller.ts (hunk 2 above)
Comment (Aherbeth) [reply api/src/app.controller.ts:40]: same as above

File: api/src/app.service.ts (hunk 3 above)
Comment (Aherbeth) [reply api/src/app.service.ts:23]: done

File: api/src/app.service.ts (hunk 4 above)
Comment (Aherbeth) [reply api/src/app.service.ts:27]: done, clearer

File: api/src/game.ts (hunk 5 above)
Comment (Aherbeth) [reply api/src/game.ts:7]: We need to modify

File: api/src/game.ts (hunk 6 above)
Comment (Aherbeth) [reply api/src/game.ts:66]: This function iterate thought the graph to check marbles in the outside nodes.
And return the value of the node found outside if there is one.
changed to `marbleColorExited`

File: api/src/app.service.ts (hunk 7 above)
Comment (Aherbeth) [reply api/src/app.service.ts:68]: Done properly

File: api/src/game.ts (hunk 8 above)
Comment (Aherbeth) [reply api/src/game.ts:121]: To ensure immutability of the state.

File: client/src/apiCalls.ts (hunk 9 above)
Comment (Aherbeth) [reply client/src/apiCalls.ts:49]: corrected

File: client/src/apiCalls.ts (hunk 10 above)
Comment (Aherbeth) [reply client/src/apiCalls.ts:65]: corrected

File: client/src/apiCalls.ts (hunk 11 above)
Comment (Aherbeth) [reply client/src/apiCalls.ts:83]: corrected

File: client/src/apiCalls.ts (hunk 12 above)
Comment (Aherbeth) [reply client/src/apiCalls.ts:103]: corrected

File: client/src/apiCalls.ts (hunk 13 above)
Comment (Aherbeth) [reply client/src/apiCalls.ts:144]: corrected

File: client/src/apiCalls.ts (hunk 14 above)
Comment (Aherbeth) [reply client/src/apiCalls.ts:159]: corrected

File: api/src/game.ts (hunk 6 above)
Comment (djhi) [reply api/src/game.ts]: `marbleColorExited` does seems clearer to me XD

File: api/src/game.ts [hunk 15]
```diff
@@ -49,13 +59,11 @@ export const switchToNextPlayer = (
...
-  let nextPlayerIndex = players.indexOf(actualPlayer) + 1;
-
-  if (nextPlayerIndex > players.length - 1) return players[0];
-  return players[nextPlayerIndex];
+  if (actualPlayer.playerNumber === 1) return players[1];
+  return players[0];
 };
 
-export const marbleWonByPlayer = (graph: Graph): number => {
+export const marbleColorExited = (graph: Graph): number => {
```
Comment (djhi) [inline api/src/game.ts:66]: As said before, `marbleColorExited` doesn't explain what this function does. I got a clue by reading the tests. Maybe `getMarbleWonByPlayer`?

Comment (Aherbeth) [reply api/src/game.ts]: Ok, i see, this is obvious. corrected

**Review by djhi — APPROVED** [review]

**Pull request merged**